include/psa/crypto_extra.h

```diff
@@ -1904,35 +1968,92 @@ static inline void psa_pake_cs_set_hash(psa_pake_cipher_suite_t *cipher_suite,
     }
 }
 
-#if defined(MBEDTLS_PSA_BUILTIN_ALG_JPAKE)
-#include <mbedtls/ecjpake.h>
-/* Note: the format for mbedtls_ecjpake_read/write function has an extra
- * length byte for each step, plus an extra 3 bytes for ECParameters in the
- * server's 2nd round. */
-#define MBEDTLS_PSA_PAKE_BUFFER_SIZE ((3 + 1 + 65 + 1 + 65 + 1 + 32) * 2)
-#endif
+struct psa_crypto_driver_pake_inputs_s {
+    uint8_t *MBEDTLS_PRIVATE(password);
+    size_t MBEDTLS_PRIVATE(password_len);
+    psa_pake_role_t MBEDTLS_PRIVATE(role);
+    psa_key_lifetime_t MBEDTLS_PRIVATE(key_lifetime);
+    psa_pake_cipher_suite_t MBEDTLS_PRIVATE(cipher_suite);
+};
 
-struct psa_pake_operation_s {
-    psa_algorithm_t MBEDTLS_PRIVATE(alg);
+enum psa_jpake_step {
+    PSA_PAKE_STEP_INVALID       = 0,
+    PSA_PAKE_STEP_X1_X2         = 1,
+    PSA_PAKE_STEP_X2S           = 2,
+    PSA_PAKE_STEP_DERIVE        = 3,
+};
+
+enum psa_jpake_state {
+    PSA_PAKE_STATE_INVALID      = 0,
+    PSA_PAKE_STATE_SETUP        = 1,
+    PSA_PAKE_STATE_READY        = 2,
+    PSA_PAKE_OUTPUT_X1_X2       = 3,
+    PSA_PAKE_OUTPUT_X2S         = 4,
+    PSA_PAKE_INPUT_X1_X2        = 5,
+    PSA_PAKE_INPUT_X4S          = 6,
+};
+
+enum psa_jpake_sequence {
+    PSA_PAKE_SEQ_INVALID        = 0,
+    PSA_PAKE_X1_STEP_KEY_SHARE  = 1,    /* also X2S & X4S KEY_SHARE */
+    PSA_PAKE_X1_STEP_ZK_PUBLIC  = 2,    /* also X2S & X4S ZK_PUBLIC */
+    PSA_PAKE_X1_STEP_ZK_PROOF   = 3,    /* also X2S & X4S ZK_PROOF */
+    PSA_PAKE_X2_STEP_KEY_SHARE  = 4,
+    PSA_PAKE_X2_STEP_ZK_PUBLIC  = 5,
+    PSA_PAKE_X2_STEP_ZK_PROOF   = 6,
+    PSA_PAKE_SEQ_END            = 7,
+};
```

Comment:
I agree that current state machine for jpake is complicated and can be improved, but I think it should be done in separate task as this PR is big enough. Potential optimization would require also updating the tests which are also complicated and I think would require another design review.
Because of this I didn't want to touch the existing sate machine and only moved it from the bult in implementation layer to core. On @silabs-hannes request added translation functions to simplify jpake step on the driver side:
```
typedef enum psa_pake_driver_step {
    PSA_JPAKE_STEP_INVALID        = 0,  /* Invalid step */
    PSA_JPAKE_X1_STEP_KEY_SHARE   = 1,  /* Round 1: input/output key share (for ephemeral private key X1).*/
    PSA_JPAKE_X1_STEP_ZK_PUBLIC   = 2,  /* Round 1: input/output Schnorr NIZKP public key for the X1 key */
    PSA_JPAKE_X1_STEP_ZK_PROOF    = 3,  /* Round 1: input/output Schnorr NIZKP proof for the X1 key */
    PSA_JPAKE_X2_STEP_KEY_SHARE   = 4,  /* Round 1: input/output key share (for ephemeral private key X2).*/
    PSA_JPAKE_X2_STEP_ZK_PUBLIC   = 5,  /* Round 1: input/output Schnorr NIZKP public key for the X2 key */
    PSA_JPAKE_X2_STEP_ZK_PROOF    = 6,  /* Round 1: input/output Schnorr NIZKP proof for the X2 key */
    PSA_JPAKE_X2S_STEP_KEY_SHARE  = 7,  /* Round 2: output X2S key (our key) */
    PSA_JPAKE_X2S_STEP_ZK_PUBLIC  = 8,  /* Round 2: output Schnorr NIZKP public key for the X2S key (our key) */
    PSA_JPAKE_X2S_STEP_ZK_PROOF   = 9,  /* Round 2: output Schnorr NIZKP proof for the X2S key (our key) */
    PSA_JPAKE_X4S_STEP_KEY_SHARE  = 10, /* Round 2: input X4S key (from peer) */
    PSA_JPAKE_X4S_STEP_ZK_PUBLIC  = 11, /* Round 2: input Schnorr NIZKP public key for the X4S key (from peer) */
    PSA_JPAKE_X4S_STEP_ZK_PROOF   = 12  /* Round 2: input Schnorr NIZKP proof for the X4S key (from peer) */
} psa_pake_driver_step_t;
```

My suggestion is to create separate tasks: for the gap in the sequence check and for optimization of the existing state machine (where the new design can be discussed). What do you think?

Reply:
I agree that this PR is a big one thus fine by me to not change the state machine in this PR. I will create a separate issue for that. Otherwise, I would actually prefer to split this PR as I foresee that I will in fact do successive and independent reviews. In this PR I would prefer to keep only the definition of the driver interface and the split of the functions in psa_crypto_pake.c to functions in psa_crypto.c and psa_crypto_pake.c: the psa_pake_*() functions in psa_crypto.c just call the  mbedtls_psa_pake_*() functions not the driver wrapper. The split would be validated by test_suite_psa_crypto_pake.
Then in follow-up PRs (I think it can be done and it would be better to do it in several PRs), we introduce the driver wrapper, the dispatch to a transparent and opaque test driver, the testing of the dispatch and then the testing of the PSA configuration for JPAKE (libtestdriver1). Wdyt?

Reply:
I think that splitting this PR into few PRs with some intermediate states (with tests adapted to this states) is complicated and will take a lot of time.
I did everything in one PR because my knowledge about drivers was very limited and I had to create all parts and then glue them together and test to see if this solution works. PR is big but a lot of code especially related to the driver dispatch logic is a copy/paste from other drivers (driver wrappers, transparent/opaque drivers, compositions, etc.) with some minor adaptations for pake. 
My suggestion is to make the implementation provided in this PR consistent with the agreed design and continue with possible improvements in the follow-up tasks.

Reply:
> I think that splitting this PR into few PRs with some intermediate states (with tests adapted to this states) is complicated and will take a lot of time.

I do not think so and I can guide you through that. We have been through the "it is big and it is difficult/time consuming to split it belief" a couple of time with driver interface and driver dispatch related PRs. I believe it is better to stop doing this. I think the goal for the first PR is clear as explained above. Stated otherwise from the current PR, we would mostly just remove the changes in the wrapper and remove the code guarded by PSA_CRYPTO_DRIVER_TEST. There would be  no additional tests, no change in tests (but for the addition of the `pake_input_getters` tests probably) as it would be just a re-organization of the current implementation around the newly defined driver interface. And to validate this re-organization, we just need the already existing tests in test_suite_psa_crypto_pake.

Reply:
Well starting over and creating new PRs with some intermediate states, step by step, will take weeks and we will end up with something like this. But I can do that if this is approach is approved by @mpg and @daverodgman.

I don't see the point with removing driver wrappers in first step. PAKE driver wrappers are similar to other driver wrappers and `test_suite_psa_crypto_pake` when run in configuration without drivers uses build in implementation anyway.

Reply:
> Well starting over and creating new PRs with some intermediate states, step by step, will take weeks ...

Again I do not think this is true. This is additional work and time but not weeks, just a couple of days: first PR, definition of the driver interface and re-organization of the implementation around it; second PR, introduction of the transparent test driver, the dispatch to it and related testing; third PR, introduction of the opaque test driver, the dispatch to it and related testing; fourth PR  testing of the PSA configuration related to PAKE (libtestdriver1). As you said, everything is there, it is just git history re-writing.

Having said that this would not really ease my review (that I will do following the above steps by the way) thus I am not going to insist on it. I hope you will consider the above if you have to work on new PSA APIs in the future.

Reply:
> I hope you will consider the above if you have to work on new PSA APIs in the future.

Actually I was going to suggest that perhaps the splitting you suggest could be documented somewhere in order to ease future work. I think Gilles already has a PR with a checklist of things to do when adding a new algorithm, perhaps it could be expanded with this?

ISTR Gilles or you suggested how some steps might be further broken down, like we even before we have a software implementation we can implement and negative-test the part of the logic that's in the core.

For various reasons the EC J-PAKE work wasn't nicely broken down before being started, I think think we can all agree that the result is less than pleasant - huge PRs that take months to get merged, issues missed during review... I fully agree we should use this as a lesson and do better in the future (I think our next opportunity will be PBKDF2).

(In case it needs clarifying: I'm not blaming anyone, especially not the people who created/reworked the PRs; investigation and break-down should have happened before PRs were created. The fact that it didn't is a shared responsibility.)